src/charm.py

```diff
@@ -249,52 +263,96 @@ def _configure(self, _):
         line arguments). If the Pebble layer has changed then Prometheus
         is restarted.
         """
-        container = self.unit.get_container(self._name)
+        early_return_statuses = {
+            "cfg_load_fail": BlockedStatus(
+                "Prometheus failed to reload the configuration (WAL replay or ingress in progress?); see debug logs"
+            ),
+            "restart_fail": BlockedStatus(
+                "Prometheus failed to restart (config valid?); see debug logs"
+            ),
+            "push_fail": BlockedStatus(
+                "Failed to push updated config/alert files; see debug logs"
+            ),
+            "layer_fail": BlockedStatus("Failed to update prometheus service; see debug logs"),
+            "config_invalid": BlockedStatus("Invalid prometheus configuration; see debug logs"),
+            "validation_fail": BlockedStatus(
+                "Failed to validate prometheus config; see debug logs"
+            ),
+        }
 
         if not self.resources_patch.is_ready():
             if isinstance(self.unit.status, ActiveStatus) or self.unit.status.message == "":
                 self.unit.status = WaitingStatus("Waiting for resource limit patch to apply")
             return
 
-        if not container.can_connect():
+        if not self.container.can_connect():
             self.unit.status = MaintenanceStatus("Configuring Prometheus")
             return
 
-        self._generate_prometheus_config(container)
-        logger.info("Pushed new configuration")
-
-        self._set_alerts(container)
+        try:
+            # Need to reload if config or alerts changed.
+            # (Both functions need to run so cannot use the short-circuiting `or`.)
+            should_reload = any(
+                [
+                    self._generate_prometheus_config(self.container),
+                    self._set_alerts(self.container),
+                ]
+            )
+        except PebbleError as e:
+            logger.error("Failed to push updated config/alert files: %s", e)
+            self.unit.status = early_return_statuses["push_fail"]
+            return
 
-        current_services = container.get_plan().services
-        new_layer = self._prometheus_layer
+        try:
+            should_restart = self._update_layer(self.container)
+        except (TypeError, PebbleError) as e:
+            logger.error("Failed to update prometheus service: %s", e)
+            self.unit.status = early_return_statuses["layer_fail"]
+            return
 
-        if current_services == new_layer.services:
-            reloaded = self._prometheus_server.reload_configuration()
-            if not reloaded:
-                logger.error("Prometheus failed to reload the configuration")
-                self.unit.status = BlockedStatus(CORRUPT_PROMETHEUS_CONFIG_MESSAGE)
+        try:
+            output, err = self._promtail_check_config()
```

Comment:
I worry that this is one of those places where unit testing won't actually approximate what happens.

Specifically, if we add a layer _and_ a config which has a config problem and fails to start, we're going to return just above this with `layer_fail`. Logically, by the time we hit this block at all, it's already succeeded if there are any layer changes, and we don't know whether the config is valid or not (which could be a good/valid reason why the layer won't start)

So, the steps are:

* `generate_config` (and push it), which is not obvious unless you read the method.
* `_update_layer` if the hash differs

It seems that we should do something like:

* `generate_config`, but don't push it yet. Write it to a temporary path, try to check it with promtool. If it fails, set the message. If it passes, move it into the right place and continue. If it fails, do as we are doing now (since checking the config will fail, and `should_restart` won't trigger the "real" restart). The current codepath has a dead zone where an invalid config has been generated and pushed to the container but actually applied yet.
  - Alternatively, generate it, push it, try to restart anyway, and if there's an exception failing to replan, validate the config there and raise this message.

Reply:
`_update_layer` only calls `container.add_layer`, so iiuc no restart should take place just yet.